README
NAME
    RT::Extension::TicketLocking - Enables users to place advisory locks on
    tickets

DESCRIPTION
    Locks can be of several different types. Current types are:

    hard (manual) lock
        A lock can be initiated manually by clicking the "Lock" link on one
        of the pages for the ticket. However, hard locks are available only
        to users who can ModifyTicket.

    take lock
        This is only applicable within RTIR. See "RTIR" section below.

    auto lock
        A lock is created whenever a user performs an action on a ticket
        that takes multiple steps if a hard lock is not already in place for
        that ticket.

        An auto lock is removed once the user is done with whatever he was
        doing on the page (e.g., when he clicks "Save Changes" on the Edit
        page). It is also removed if the Unlock link is clicked from a page
        that generated an auto lock.

        Auto-lock is set for the following actions in RT:

            - Comment
            - Reply
            - Resolve

        RTIR's user may find list of actions below.

    Locks are advisory: if a ticket is locked by one user, other users will
    be given a notification (in red) that another user has locked the
    ticket, with the locking user's name and how long he has had it locked
    for, but they will still be allowed to edit and submit changes on the
    ticket.

    When a user locks a ticket (auto lock or hard lock), they are given a
    notification informing them of their lock and how long they have had the
    ticket locked (in some other color - currently green).

  Removing locks
    Locks will remain in place until:

    *   The user is done editing/replying/etc. (for auto locks, if there is
        no hard lock on the ticket)

    *   A lock can be removed manually by clicking the "Unlock" link on one
        of the pages for the ticket. This removes any type of lock.

    *   The user logs out

    *   A configurable expiry period has elapsed (if the $LockExpiry config
        variable has been set to a value greater than zero)

    When a user unlocks a ticket (auto unlock or hard unlock), they are
    given a notification informing them that their lock has been removed,
    and how long they had the ticket locked for.

  Merging tickets
    When a locked ticket (hard or take lock) is merged into another ticket,
    the ticket being merged into will get the lock type of the ticket being
    merged from. This lock shift is conditional upon priority, as usual - if
    the merged from ticket has a lock of a lower priority than the merged-to
    ticket, the merged-to ticket will retain its lock. If the merged-to
    ticket is locked by a different user, that user will retain the lock.
    Basically, the merged-to ticket will retain its lock if it is higher
    priority than the lock on the ticket being merged from.

  RTIR
    Within RTIR auto locks are applied for the following actions:

        - Edit
        - Split
        - Merge
        - Advanced
        - Reply
        - Resolve
        - Reject
        - Comment
        - Remove

    As well, there is special type of lock implemented in RTIR. When a user
    clicks the "Take" link for an RTIR Incident ticket, a Take lock is
    added. This lock will only be removed when the IR is linked to a new or
    existing Incident. If RTIR is not installed, this type will not be
    available.

INSTALLATION
    perl Makefile.PL
    make
    make install
        May need root permissions

    Edit your /opt/rt4/etc/RT_SiteConfig.pm
        If you are using RT 4.2 or greater, add this line:

            Plugin('RT::Extension::TicketLocking');

        For RT 4.0, add this line:

            Set(@Plugins, qw(RT::Extension::TicketLocking));

        or add RT::Extension::TicketLocking to your existing @Plugins line.

    Clear your mason cache
            rm -rf /opt/rt4/var/mason_data/obj

    Restart your webserver

CONFIGURATION
  LockExpiry option
    In the config you can set LockExpiry option to a number of seconds, the
    longest time a lock can remain without being automatically removed, for
    example:

        Set( $LockExpiry, 5*60 ); # lock expires after five minutes

    If you don't wish to have your locks automatically expire, simply set
    $LockExpiry to a false (zero or undef) value. This is the default if you
    do not provide a $LockExpiry.

  Allowing users to use 'MyLocks' portlet
    The extension comes with a portlet users can place on thier home page
    RT's or RTIR's. Using this portlet user can easily jump to locked
    tickets, remove particular lock or all locks at once.

    If you want the MyLocks portlet to be available then you have to place
    it in the list of allowed components.

    For RT:

        Set($HomepageComponents, [qw(
            MyLocks 
            ... list of another portlets ...
        )]);

    People can then choose to add the portlet to their homepage in
    Preferences -> 'RT at a glance'.

    If you are running RTIR, and want the portlet to be available from the
    RTIR home page, you will need to do something similar to set the
    RTIR_HomepageComponents array in your config file, like this:

        Set(@RTIR_HomepageComponents, qw(
            MyLocks
            ... list of another portlets ...
        ));

    People can then choose to add the portlet to their homepage in
    Preferences -> 'RTIR Home'.

AUTHOR
    Best Practical Solutions, LLC <modules@bestpractical.com>

BUGS
    All bugs should be reported via email to

        L<bug-RT-Extension-TicketLocking@rt.cpan.org|mailto:bug-RT-Extension-TicketLocking@rt.cpan.org>

    or via the web at

        L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-TicketLocking>.

COPYRIGHT
    This extension is Copyright (C) 2007-2014 Best Practical Solutions, LLC.

    This is free software, licensed under:

      The GNU General Public License, Version 2, June 1991



Changes
Revision history for RT-Extension-TicketLocking

1.08 2022-03-31
 - Update code for RT 5

1.07 2018-02-15

 - update code for RTIR 3.0+
 - update MyLocks component for RT 4.4

1.06 2016-10-21
 - Update documentation

1.05 2015-02-26
 - Packaging updates
 - Localization, thanks to Emmanuel Lacour

1.00 2015-01-19
 - Fix escaping in "locked by..."
 - Bump dependency to 4.0 that contains l_unsafe
 - Updated packaging

0.12 2013-10-10
 - port menu items over RT 4.0+

0.11 2013-10-08
 - update installer
 - use ShowUser component in the message
 - styling clean ups for MyLocks

0.10 2008-09-15
 - update installer

0.09 2007-03-05
 - added this file
 - updated some metadata


t/test_suite.pl
#!/usr/bin/perl

# Load this in test scripts with: require "t/test_suite.pl";
# *AFTER* loading in Test::More.


use strict;
use warnings;

use HTTP::Cookies;
use Test::More;

### after: use lib qw(@RT_LIB_PATH@);
use lib qw(/opt/rt3/local/lib /opt/rt3/lib);

use RT;
RT::LoadConfig();
RT::Init();

my $RTIR_TEST_USER = "rtir_test_user";
my $RTIR_TEST_PASS = "rtir_test_pass";

require RT::Test;
use RT::Test::Web;

sub default_agent {
    my $agent = new RT::Test::Web;
    $agent->cookie_jar( HTTP::Cookies->new );
    my $u = rtir_user();
    $agent->login($RTIR_TEST_USER, $RTIR_TEST_PASS);
    $agent->get_ok("/index.html", "loaded home page");
    return $agent;
}

sub set_custom_field {
    my $agent = shift;
    my $cf_name = shift;
    my $val = shift;
    my $field_name = $agent->value($cf_name) or return 0;
    $agent->field($field_name, $val);
    return 1;
}

sub display_ticket {
    my $agent = shift;
    my $id = shift;

    $agent->get_ok("/RTIR/Display.html?id=$id", "Loaded Display page for Ticket #$id");
}

sub ticket_state_is {
    my $agent = shift;
    my $id = shift;
    my $state = shift;
    my $desc = shift || "State of the ticket #$id is '$state'";
    display_ticket( $agent, $id );
    $agent->content =~ qr{State:\s*</td>\s*<td[^>]*?>\s*<span class="cf-value">([\w ]+)</span>}ism;
    return is($1, $state, $desc);
}


sub create_user {
    my $user_obj = rtir_user();

    ok($user_obj->Id > 0, "Successfully found the user");
    
    my $group_obj = RT::Group->new(RT::SystemUser());
    $group_obj->LoadUserDefinedGroup("DutyTeam");
    ok($group_obj->Id > 0, "Successfully found the DutyTeam group");

    $group_obj->AddMember($user_obj->Id);
    ok($group_obj->HasMember($user_obj->PrincipalObj), "user is in the group");
}

sub rtir_user {
    my $u = RT::Test->load_or_create_user(
        Name         => $RTIR_TEST_USER,
        Password     => $RTIR_TEST_PASS,
        EmailAddress => "$RTIR_TEST_USER\@example.com",
        RealName     => "$RTIR_TEST_USER Smith",
        Privileged   => 1,
    );
    return $u;
}

sub create_incident {
    return create_rtir_ticket_ok( shift, 'Incidents', @_ );
}
sub create_ir {
    return create_rtir_ticket_ok( shift, 'Incident Reports', @_ );
}
sub create_investigation {
    return create_rtir_ticket_ok( shift, 'Investigations', @_ );
}
sub create_block {
    return create_rtir_ticket_ok( shift, 'Blocks', @_ );
}

sub goto_create_rtir_ticket {
    my $agent = shift;
    my $queue = shift;

    my %type = (
        'Incident Reports' => 'Report',
        'Investigations'   => 'Investigation',
        'Blocks'           => 'Block',
        'Incidents'        => 'Incident'
    );

    $agent->get_ok("/RTIR/index.html", "loaded home page");

    $agent->follow_link_ok({text => $queue, n => "1"}, "Followed '$queue' link");
    $agent->follow_link_ok({text => "New ". $type{ $queue }, n => "1"}, "Followed 'New $type{$queue}' link");
    

    # set the form
    $agent->form_number(3);
}

sub create_rtir_ticket_ok {
    my $agent = shift;
    my $queue = shift;

    my $id = create_rtir_ticket( $agent, $queue, @_ );
    ok $id, "Created ticket #$id in queue '$queue' successfully.";
    return $id;
}

sub create_rtir_ticket
{
    my $agent = shift;
    my $queue = shift;
    my $fields = shift || {};
    my $cfs = shift || {};

    goto_create_rtir_ticket($agent, $queue);
    
    #Enable test scripts to pass in the name of the owner rather than the ID
    if ($$fields{Owner} && $$fields{Owner} !~ /^\d+$/)
    {
        if($agent->content =~ qr{<option.+?value="(\d+)"\s*>$$fields{Owner}</option>}ims) {
            $$fields{Owner} = $1;
        }
    }
    

    $fields->{'Requestors'} ||= $RTIR_TEST_USER if $queue eq 'Investigations';
    while (my ($f, $v) = each %$fields) {
        $agent->field($f, $v);
    }

    while (my ($f, $v) = each %$cfs) {
        set_custom_field($agent, $f, $v);
    }

    my %create = (
        'Incident Reports' => 'Create',
        'Investigations'   => 'Create',
        'Blocks'           => 'Create',
        'Incidents'        => 'CreateIncident'
    );
    # Create it!
    $agent->click( $create{ $queue } );
    
    is ($agent->status, 200, "Attempted to create the ticket");

    return get_ticket_id($agent);
}

sub get_ticket_id {
    my $agent = shift;
    my $content = $agent->content();
    my $id = 0;
    if ($content =~ /.*Ticket (\d+) created.*/g) {
        $id = $1;
    }
    elsif ($content =~ /.*No permission to view newly created ticket #(\d+).*/g) {
        diag "No permissions to view the ticket" if $ENV{'TEST_VERBOSE'};
    }
    else {
        diag "Couldn't find ticket id in:\n$content" if $ENV{'TEST_VERBOSE'};
    }
    return $id;
}


sub create_incident_for_ir {
    my $agent = shift;
    my $ir_id = shift;
    my $fields = shift || {};
    my $cfs = shift || {};

    display_ticket($agent, $ir_id);

    # Select the "New" link from the Display page
    $agent->follow_link_ok({text => "[New]"}, "Followed 'New (Incident)' link");

    $agent->form_number(3);

    while (my ($f, $v) = each %$fields) {
        $agent->field($f, $v);
    }

    while (my ($f, $v) = each %$cfs) {
        set_custom_field($agent, $f, $v);
    }

    $agent->click("CreateIncident");
    
    is ($agent->status, 200, "Attempting to create new incident linked to child $ir_id");

    ok ($agent->content =~ /.*Ticket (\d+) created in queue.*/g, "Incident created from child $ir_id.");
    my $incident_id = $1;

#    diag("incident ID is $incident_id");
    return $incident_id;
}

sub ok_and_content_like {
    my $agent = shift;
    my $re = shift;
    my $desc = shift || "looks good";
    
    is($agent->status, 200, "request successful");
    #like($agent->content, $re, $desc);
    $agent->content_like($re, $desc);
}


sub create_incident_and_investigation {
    my $agent = shift;
    my $fields = shift || {};
    my $cfs = shift || {};
    my $ir_id = shift;

    $ir_id ? display_ticket($agent, $ir_id)
        : $agent->get_ok("/index.html", "loaded home page");

    if($ir_id) {
        # Select the "New" link from the Display page
        $agent->follow_link_ok({text => "[New]"}, "Followed 'New (Incident)' link");
    }
    else 
    {
        $agent->follow_link_ok({text => "Incidents"}, "Followed 'Incidents' link");
        $agent->follow_link_ok({text => "New Incident", n => '1'}, "Followed 'New Incident' link");
    }

    # Fill out forms
    $agent->form_number(3);

    while (my ($f, $v) = each %$fields) {
        $agent->field($f, $v);
    }

    while (my ($f, $v) = each %$cfs) {
        set_custom_field($agent, $f, $v);
    }
    $agent->click("CreateWithInvestigation");
    my $msg = $ir_id
        ? "Attempting to create new incident and investigation linked to child $ir_id"
        : "Attempting to create new incident and investigation";
    is ($agent->status, 200, $msg);
    $msg = $ir_id ? "Incident created from child $ir_id." : "Incident created.";

    my $re = qr/.*Ticket (\d+) created in queue &#39;Incidents&#39;/;
    $agent->content_like( $re, $msg );
      my ($incident_id) = ($agent->content =~ $re);
      
    $re = qr/.*Ticket (\d+) created in queue &#39;Investigations&#39;/;
    $agent->content_like( $re, "Investigation created for Incident $incident_id." );
    my ($investigation_id) = ($agent->content =~ $re);

    return ($incident_id, $investigation_id);
}


sub create_ticket {
    my $agent = shift;
    my $queue = shift || 'General';

    return create_rtir_ticket($agent, $queue, @_) 
        if $queue eq 'Incidents'
        || $queue eq 'Blocks'
        || $queue eq 'Investigations'
        || $queue eq 'Incident Reports';
    
    my $fields = shift || {};
    my $cfs = shift || {};

    $agent->get_ok(
        "/Ticket/Create.html?Queue=$queue",
        "Went to Create page in queue $queue",
    );

    #Enable test scripts to pass in the name of the owner rather than the ID
    if ( $fields->{'Owner'} && $fields->{'Owner'} !~ /^\d+$/ ) {
        my $u = RT::User->new( $RT::SystemUser );
        $u->Load( $fields->{'Owner'} );
        die "Couldn't load user '". $fields->{'Owner'} ."'"
            unless $u->id;
        $fields->{'Owner'} = $u->id;
    }
    
    $agent->form_number(3);
    while (my ($f, $v) = each %$fields) {
        $agent->field($f, $v);
    }

    while (my ($f, $v) = each %$cfs) {
        set_custom_field($agent, $f, $v);
    }
    
    
    # Create it!
    $agent->click_button(value => 'Create');
    
    is ($agent->status, 200, "Attempted to create the ticket");

    return get_ticket_id($agent);
}

1;


xt/test_suite.pl
#!/usr/bin/perl

use strict;
use warnings;

use HTTP::Cookies;
require RT::IR::Test::Web;

### after: use lib qw(@RT_LIB_PATH@);
use lib qw(/opt/rt4/local/lib /opt/rt4/lib);

my $RTIR_TEST_USER = "rtir_test_user";
my $RTIR_TEST_PASS = "rtir_test_pass";

sub default_agent {
    my $agent = new RT::Test::Web;
    $agent->cookie_jar( HTTP::Cookies->new );
    my $u = rtir_user();
    $agent->login($RTIR_TEST_USER, $RTIR_TEST_PASS);
    $agent->get_ok("/index.html", "loaded home page");
    return $agent;
}

sub default_rtir_agent {
    my $agent = new RT::IR::Test::Web;
    $agent->cookie_jar( HTTP::Cookies->new );
    my $u = rtir_user(MemberOf => 'DutyTeam');
    $agent->login($RTIR_TEST_USER, $RTIR_TEST_PASS);
    $agent->get_ok("/index.html", "loaded home page");
    return $agent;
}

sub rtir_user {
    my $u = RT::Test->load_or_create_user(
        Name         => $RTIR_TEST_USER,
        Password     => $RTIR_TEST_PASS,
        EmailAddress => "$RTIR_TEST_USER\@example.com",
        RealName     => "$RTIR_TEST_USER Smith",
        Privileged   => 1,
        @_,
    );
    return $u;
}


sub create_ticket {
    my $agent = shift;
    my $queue = shift || 'General';

    my $fields = shift || {};
    my $cfs = shift || {};

    my $q = RT::Test->load_or_create_queue(Name => $queue);

    $agent->goto_create_ticket($q);

    #Enable test scripts to pass in the name of the owner rather than the ID
    if ( $fields->{'Owner'} && $fields->{'Owner'} !~ /^\d+$/ ) {
        my $u = RT::User->new( $RT::SystemUser );
        $u->Load( $fields->{'Owner'} );
        die "Couldn't load user '". $fields->{'Owner'} ."'"
            unless $u->id;
        $fields->{'Owner'} = $u->id;
    }
    
    $agent->form_number(3);
    while (my ($f, $v) = each %$fields) {
        $agent->field($f, $v);
    }

    while (my ($f, $v) = each %$cfs) {
        set_custom_field($agent, $f, $v);
    }
    
    
    # Create it!
    $agent->click_button(value => 'Create');
    
    is ($agent->status, 200, "Attempted to create the ticket");

    return get_ticket_id($agent);
}

1;


lib/RT/Extension/TicketLocking.pm
# BEGIN BPS TAGGED BLOCK {{{
#
# COPYRIGHT:
#
# This software is Copyright (c) 2007-2015 Best Practical Solutions, LLC
#                                          <sales@bestpractical.com>
#
# (Except where explicitly superseded by other copyright notices)
#
#
# LICENSE:
#
# This work is made available to you under the terms of Version 2 of
# the GNU General Public License. A copy of that license should have
# been provided with this software, but in any event can be snarfed
# from www.gnu.org.
#
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 or visit their web page on the internet at
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.html.
#
#
# CONTRIBUTION SUBMISSION POLICY:
#
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of
# the GNU General Public License and is only of importance to you if
# you choose to contribute your changes and enhancements to the
# community by submitting them to Best Practical Solutions, LLC.)
#
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with
# Request Tracker, to Best Practical Solutions, LLC, you confirm that
# you are the copyright holder for those contributions and you grant
# Best Practical Solutions,  LLC a nonexclusive, worldwide, irrevocable,
# royalty-free, perpetual, license to use, copy, create derivative
# works based on those contributions, and sublicense and distribute
# those contributions and any derivatives thereof.
#
# END BPS TAGGED BLOCK }}}

use v5.8.3;
use strict;
use warnings;

package RT::Extension::TicketLocking;

our $VERSION = '1.08';

=head1 NAME

RT::Extension::TicketLocking - Enables users to place advisory locks on tickets

=head1 RT VERSION

Works with RT 4 and 5.0.

=head1 DESCRIPTION

Locks can be of several different types. Current types are:

=over 4

=item hard (manual) lock

A lock can be initiated manually by clicking the "Lock" link on one of the pages
for the ticket. However, hard locks are available only to users who can ModifyTicket.

=item take lock

This is only applicable within RTIR. See L</RTIR> section below.

=item auto lock

A lock is created whenever a user performs an action on a ticket that takes
multiple steps if a hard lock is not already in place for that ticket.

An auto lock is removed once the user is done with whatever he was doing
on the page (e.g., when he clicks "Save Changes" on the Edit page).
It is also removed if the Unlock link is clicked from a page that generated
an auto lock.

Auto-lock is set for the following actions in RT:

    - Comment
    - Reply
    - Resolve

RTIR's user may find list of actions below.

=back

Locks are advisory: if a ticket is locked by one user, other users
will be given a notification (in red) that another user has locked
the ticket, with the locking user's name and how long he has had
it locked for, but they will still be allowed to edit and submit
changes on the ticket.

When a user locks a ticket (auto lock or hard lock), they are given
a notification informing them of their lock and how long they have
had the ticket locked (in some other color - currently green).

=head2 Removing locks

Locks will remain in place until:

=over 4

=item * The user is done editing/replying/etc. (for auto locks, if
there is no hard lock on the ticket)

=item * A lock can be removed manually by clicking the "Unlock" link on one
of the pages for the ticket. This removes B<any> type of lock.

=item * The user logs out

=item * A configurable expiry period has elapsed (if the $LockExpiry
config variable has been set to a value greater than zero)

=back

When a user unlocks a ticket (auto unlock or hard unlock),
they are given a notification informing them that their
lock has been removed, and how long they had the ticket
locked for.

=head2 Merging tickets

When a locked ticket (hard or take lock) is merged into another ticket,
the ticket being merged into will get the lock type of the ticket being
merged from. This lock shift is conditional upon priority, as
usual - if the merged from ticket has a lock of a lower priority than
the merged-to ticket, the merged-to ticket will retain its lock.
If the merged-to ticket is locked by a different user, that user will
retain the lock. Basically, the merged-to ticket will retain its lock
if it is higher priority than the lock on the ticket being merged from.

=head2 RTIR

Within RTIR auto locks are applied for the following actions:

    - Edit
    - Split
    - Merge
    - Advanced
    - Reply
    - Resolve
    - Reject
    - Comment
    - Remove

As well, there is special type of lock implemented in RTIR. When a
user clicks the "Take" link for an RTIR Incident ticket, a Take lock
is added. This lock will only be removed when the IR is linked to
a new or existing Incident. If RTIR is not installed, this type
will not be available.

=head1 INSTALLATION

=over

=item C<perl Makefile.PL>

=item C<make>

=item C<make install>

May need root permissions

=item Edit your F</opt/rt5/etc/RT_SiteConfig.pm>

If you are using RT 4.2 or greater, add this line:

    Plugin('RT::Extension::TicketLocking');

For RT 4.0, add this line:

    Set(@Plugins, qw(RT::Extension::TicketLocking));

or add C<RT::Extension::TicketLocking> to your existing C<@Plugins> line.

=item Clear your mason cache

    rm -rf /opt/rt5/var/mason_data/obj

=item Restart your webserver

=back


=head1 CONFIGURATION

=head2 LockExpiry option

In the config you can set LockExpiry option to a number of seconds,
the longest time a lock can remain without being automatically removed,
for example:

    Set( $LockExpiry, 5*60 ); # lock expires after five minutes

If you don't wish to have your locks automatically expire, simply
set $LockExpiry to a false (zero or undef) value. This is the default if
you do not provide a $LockExpiry.

=head2 Allowing users to use 'MyLocks' portlet

The extension comes with a portlet users can place on thier home
page RT's or RTIR's. Using this portlet user can easily jump to
locked tickets, remove particular lock or all locks at once.

If you want the MyLocks portlet to be available then you have
to place it in the list of allowed components.

For RT:

    Set($HomepageComponents, [qw(
        MyLocks 
        ... list of another portlets ...
    )]);

People can then choose to add the portlet to their homepage
in Preferences -> 'RT at a glance'.

If you are running RTIR, and want the portlet to be available
from the RTIR home page, you will need to do something similar
to set the RTIR_HomepageComponents array in your config file,
like this:

    Set(@RTIR_HomepageComponents, qw(
        MyLocks
        ... list of another portlets ...
    ));

People can then choose to add the portlet to their homepage
in Preferences -> 'RTIR Home'.

=head1 AUTHOR

Best Practical Solutions, LLC E<lt>modules@bestpractical.comE<gt>

=head1 BUGS

All bugs should be reported via email to

    L<bug-RT-Extension-TicketLocking@rt.cpan.org|mailto:bug-RT-Extension-TicketLocking@rt.cpan.org>

or via the web at

    L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-TicketLocking>.

=head1 COPYRIGHT

This extension is Copyright (C) 2007-2014 Best Practical Solutions, LLC.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut

# IMPLEMENTATION DETAILS
#  Each type is associated with a priority. Current priorities are as follows,
#  from highest priority to lowest:
#      - Hard
#      - Take (when applicable)
#      - Auto
#  
#  This allow us to store only one lock record with higher priority.

use RT::Ticket;
package RT::Ticket;

our @LockTypes = qw(Auto Take Hard);
our %CheckRightOnLock = (
    Hard => 'ModifyTicket',
);

sub LockPriority {
    my $self = shift;
    my $type = shift;
    
    my $priority;
    for( my $i = 0; $i < scalar @LockTypes; $i++) {
        $priority = $i if lc( $LockTypes[ $i ] ) eq lc( $type );
    }
    $RT::Logger->error( "There is no type '$type' in the list of lock types")
        unless defined $priority;

    return $priority || 0;
}

sub Locked {
    my $ticket = shift;

    my $lock = $ticket->FirstAttribute('RT_Lock');
    return $lock unless $lock;

    return $lock unless my $expiry = RT->Config->Get('LockExpiry');

    my $duration = time() - $lock->Content->{'Timestamp'};
    unless ( $duration < $expiry ) {
        $ticket->DeleteAttribute('RT_Lock');
        undef $lock;
    }
    return $lock;
}

sub Lock {
    my $ticket = shift;
    my $type = shift || 'Auto';

    if ( my $lock = $ticket->Locked() ) {
        return undef if $lock->Content->{'User'} != $ticket->CurrentUser->id;
        my $current_type = $lock->Content->{'Type'};
        return undef if $ticket->LockPriority( $type ) <= $ticket->LockPriority( $current_type );
    }

    if ( my $right = $CheckRightOnLock{ $type } ) {
        return undef unless $ticket->CurrentUserHasRight('ModifyTicket');
    }

    $ticket->Unlock($type);    #Remove any existing locks (because this one has greater priority)
    my $id = $ticket->id;
    my $username = $ticket->CurrentUser->Name;
    $ticket->SetAttribute(
        Name    => 'RT_Lock',
        Description => "$type lock on Ticket $id by user $username",
        Content => {
            User      => $ticket->CurrentUser->id,
            Timestamp => time(),
            Type => $type,
            Ticket => $id
        }
    );
}


sub Unlock {
    my $ticket = shift;
    my $type = shift || 'Auto';

    my $lock = $ticket->RT::Ticket::Locked();
    return (undef, $ticket->CurrentUser->loc("This ticket was not locked.")) unless $lock;
    return (undef, $ticket->CurrentUser->loc("You cannot unlock a ticket locked by another user."))
        unless $lock->Content->{User} == $ticket->CurrentUser->id;

    my $current_type = $lock->Content->{'Type'};
    return (undef, $ticket->CurrentUser->loc("There is a lock with a higher priority on this ticket."))
        if $ticket->LockPriority( $type ) < $ticket->LockPriority( $current_type );

    my $duration = time() - $lock->Content->{'Timestamp'};
    $ticket->DeleteAttribute('RT_Lock');
    return ($duration, $ticket->CurrentUser->loc("You have unlocked this ticket. It was locked for [_1] seconds.", $duration));
}


sub BreakLock {
    my $ticket = shift;
    return $ticket->DeleteAttribute('RT_Lock');
}


use RT::User;
package RT::User;

sub GetLocks {
    my $self = shift;
    
    my $attribs = RT::Attributes->new($self);
    $attribs->Limit(FIELD => 'Creator', OPERATOR=> '=', VALUE => $self->id(), ENTRYAGGREGATOR => 'AND');
    
    my $expiry = RT->Config->Get('LockExpiry');
    return $attribs->Named('RT_Lock') unless $expiry;
    my @locks;
    
    foreach my $lock ($attribs->Named('RT_Lock')) {
        my $duration = time() - $lock->Content->{'Timestamp'};
        if($duration < $expiry) {
            push @locks, $lock;
        }
        else {
            $lock->Delete();
        }
    }
    return @locks;
}

sub RemoveLocks {
    my $self = shift;
    
    my $attribs = RT::Attributes->new($self);
    $attribs->Limit(FIELD => 'Creator', OPERATOR=> '=', VALUE => $self->id(), ENTRYAGGREGATOR => 'AND');
    my @attributes = $attribs->Named('RT_Lock');
    foreach my $lock (@attributes) {
        $lock->Delete();
    }
}

1;
